File: platform/ground/planner/src/scripted.ts
/* ============================================================================
 * eis-planner/scripted — deterministic no-network planner.
 *
 * The default demo/test path (EIS_PLANNER_MODE=scripted) uses this planner;
 * the live LLM planner (llm.ts) is only constructed when explicitly enabled.
 *
 * ALL coordinates are DERIVED from the loaded site model + anomaly — nothing
 * is hardcoded, so a different site JSON produces different plans.
 * ========================================================================== */

import { Anomaly, MissionPlan } from './contract';
import { LatLon, SiteModel, haversineMeters, polygonCentroid } from './site';

/** Orbit radius used by the passing plan, meters (well above the 3 m floor). */
export const ORBIT_RADIUS_M = 25;
/** The scripted plan uses the inspect/standard profile, whose fixture cap is 45 m AGL. */
export const SCRIPTED_PROFILE_MAX_ALT_M = 45;

/** How far above the alt band's max the failing plan flies, meters. */
export const FAILING_ALT_OVERSHOOT_M = 10;

/** How far below the alt band's min the failing plan flies when the first
 *  NFZ's ceiling is inside the band (so an above-band altitude could not also
 *  violate the NFZ), meters. */
export const FAILING_ALT_UNDERSHOOT_M = 5;

/** Point on the observation ring facing home, so approach never breaches standoff. */
export function orbitApproachPoint(home: LatLon, center: LatLon, radiusM: number): LatLon {
  const distance = haversineMeters(home, center);
  if (distance < 0.01) return { lat: center.lat + radiusM / 111_320, lon: center.lon };
  const scale = radiusM / distance;
  return { lat: center.lat + (home.lat - center.lat) * scale,
    lon: center.lon + (home.lon - center.lon) * scale };
}

export class ScriptedPlanner {
  /**
   * Emit a plan for the anomaly. `failing: true` returns the deliberately
   * invalid demo plan (see failingPlan), otherwise the passing plan.
   */
  plan(site: SiteModel, anomaly: Anomaly, opts: { failing?: boolean } = {}): MissionPlan {
    return opts.failing ? this.failingPlan(site, anomaly) : this.passingPlan(site, anomaly);
  }

  /**
   * A plan the verifier should PASS: goto the anomaly location at the middle
   * of the site alt band with profile 'standard', orbit it at ORBIT_RADIUS_M,
   * then RTL. (Assumes the anomaly lies inside the perimeter and outside any
   * NFZ — a mid-site anomaly; otherwise the verifier corrects/rejects it,
   * which is the verifier's job, not this planner's.)
   */
  passingPlan(site: SiteModel, anomaly: Anomaly): MissionPlan {
    const midAlt = Math.min(
      (site.altBandM.min + site.altBandM.max) / 2,
      SCRIPTED_PROFILE_MAX_ALT_M,
    );
    const approach = orbitApproachPoint(site.home, anomaly, ORBIT_RADIUS_M);
    return {
      requestId: `scripted-${anomaly.id}`,
      anomalyId: anomaly.id,
      profile: 'standard',
      rationale:
        `Scripted survey of anomaly ${anomaly.id} (${anomaly.type}, ` +
        `confidence ${anomaly.confidence}): fly to the flagged location at ` +
        `mid-band altitude ${midAlt} m AGL, orbit at ${ORBIT_RADIUS_M} m for ` +
        `observation, then return to launch.`,
      tools: [
        { tool: 'goto_gps', lat: approach.lat, lon: approach.lon, alt: midAlt, profile: 'standard' },
        { tool: 'orbit_point', lat: anomaly.lat, lon: anomaly.lon, radius: ORBIT_RADIUS_M },
        { tool: 'rtl' },
      ],
    };
  }

  /**
   * A plan the verifier must FAIL on BOTH the nfz and altitude checks (per
   * spec): a goto straight into the site's first NFZ at an altitude that is
   * outside the alt band yet still at/below the NFZ ceiling.
   *
   * Altitude choice, derived from the site:
   *  - preferred: alt_band.max + FAILING_ALT_OVERSHOOT_M (above the band),
   *    capped at the NFZ ceiling so the NFZ check also fails;
   *  - if the ceiling is not above the band max (an above-band altitude
   *    could never be inside the NFZ), fly BELOW the band instead:
   *    max(1, alt_band.min - FAILING_ALT_UNDERSHOOT_M), capped at the ceiling.
   *
   * Throws if the site declares no NFZs (there is nothing to violate).
   */
  failingPlan(site: SiteModel, anomaly: Anomaly): MissionPlan {
    if (site.nfz.length === 0) {
      throw new Error('ScriptedPlanner.failingPlan: site has no NFZs to violate');
    }
    const zone = site.nfz[0];
    const target = polygonCentroid(zone.polygon);
    const band = site.altBandM;

    let alt: number;
    if (zone.ceilingM > band.max) {
      alt = Math.min(zone.ceilingM, band.max + FAILING_ALT_OVERSHOOT_M);
    } else {
      alt = Math.min(zone.ceilingM, Math.max(1, band.min - FAILING_ALT_UNDERSHOOT_M));
    }

    return {
      requestId: `scripted-failing-${anomaly.id}`,
      anomalyId: anomaly.id,
      profile: 'standard',
      rationale:
        `Deliberately invalid demo plan for anomaly ${anomaly.id}: flies into ` +
        `NFZ "${zone.name}" at ${alt} m AGL, which is both inside the no-fly ` +
        `ceiling (${zone.ceilingM} m) and outside the site alt band ` +
        `[${band.min}, ${band.max}] m. The MissionVerifier must flag it.`,
      tools: [
        { tool: 'goto_gps', lat: target.lat, lon: target.lon, alt, profile: 'standard' },
        { tool: 'rtl' },
      ],
    };
  }
}

File: rails/scripted.py
"""
rails/scripted — the ``EIS_TEST_BAD_PLAN`` rail, and nothing else.

Reference of ``platform/ground/planner/src/scripted.ts``.

READ THIS BEFORE USING IT. ``ScriptedPlanner`` is NOT a plan source. The
deterministic planner (:mod:`rails.deterministic`) is the only source of flight
plans — ADR D20 — and this class survives for exactly one purpose: the
``EIS_TEST_BAD_PLAN`` demo rail, where a deliberately invalid plan is injected
so an audience can watch the trust layer bounce it. Nothing on the demo or test
flight path may call :meth:`ScriptedPlanner.passing_plan` as a planner.

All coordinates are DERIVED from the loaded site model and the anomaly, so a
different site JSON produces different plans; nothing is hardcoded.
"""
from __future__ import annotations

from .geometry import Point, haversine_meters, polygon_centroid
from .schemas import Anomaly, MissionPlan
from .site import SiteModel

#: Orbit radius used by the passing plan, metres (well above the 3 m floor).
ORBIT_RADIUS_M = 25.0
#: The scripted plan uses the inspect/standard profile, capped at 45 m AGL.
SCRIPTED_PROFILE_MAX_ALT_M = 45.0
#: How far above the alt band's max the failing plan flies, metres.
FAILING_ALT_OVERSHOOT_M = 10.0
#: How far below the alt band's min the failing plan flies when the first NFZ's
#: ceiling sits inside the band (so an above-band altitude could not also
#: violate the NFZ), metres.
FAILING_ALT_UNDERSHOOT_M = 5.0


def orbit_approach_point(home: Point, center: Point, radius_m: float) -> Point:
    """Point on the observation ring facing home, so approach never breaches standoff."""
    distance = haversine_meters(home, center)
    if distance < 0.01:
        return (center[0] + radius_m / 111_320.0, center[1])
    scale = radius_m / distance
    return (center[0] + (home[0] - center[0]) * scale,
            center[1] + (home[1] - center[1]) * scale)


class ScriptedPlanner:
    """The EIS_TEST_BAD_PLAN rail. Never a plan source on a flight path."""

    def plan(self, site: SiteModel, anomaly: Anomaly, *, failing: bool = False) -> MissionPlan:
        return self.failing_plan(site, anomaly) if failing else self.passing_plan(site, anomaly)

    def passing_plan(self, site: SiteModel, anomaly: Anomaly) -> MissionPlan:
        """A plan the verifier should PASS, for the demo's "before" frame."""
        mid_alt = min((site.altBandMin + site.altBandMax) / 2.0, SCRIPTED_PROFILE_MAX_ALT_M)
        home: Point = (site.home.lat, site.home.lon)
        approach = orbit_approach_point(home, (anomaly.lat, anomaly.lon), ORBIT_RADIUS_M)
        return MissionPlan(
            requestId=f"scripted-{anomaly.id}",
            anomalyId=anomaly.id,
            profile="standard",
            rationale=(
                f"Scripted survey of anomaly {anomaly.id} ({anomaly.type}, confidence "
                f"{anomaly.confidence}): fly to the flagged location at mid-band altitude "
                f"{mid_alt} m AGL, orbit at {ORBIT_RADIUS_M} m for observation, then return to launch."
            ),
            tools=[
                {"tool": "goto_gps", "lat": approach[0], "lon": approach[1],
                 "alt": mid_alt, "profile": "standard"},
                {"tool": "orbit_point", "lat": anomaly.lat, "lon": anomaly.lon,
                 "radius": ORBIT_RADIUS_M},
                {"tool": "rtl"},
            ],
        )

    def failing_plan(self, site: SiteModel, anomaly: Anomaly) -> MissionPlan:
        """The deliberately invalid plan: into an NFZ, outside the altitude band.

        Fails BOTH the nfz and the altitude check. The altitude is derived from
        the site: preferably above the band but still at/below the NFZ ceiling;
        when the ceiling is not above the band max (so an above-band altitude
        could never be inside the NFZ), below the band instead.
        """
        if not site.nfz:
            raise ValueError("ScriptedPlanner.failing_plan: site has no NFZs to violate")
        zone = site.nfz[0]
        target = polygon_centroid(zone.polygon)
        if zone.ceilingM > site.altBandMax:
            alt = min(zone.ceilingM, site.altBandMax + FAILING_ALT_OVERSHOOT_M)
        else:
            alt = min(zone.ceilingM, max(1.0, site.altBandMin - FAILING_ALT_UNDERSHOOT_M))
        return MissionPlan(
            requestId=f"scripted-failing-{anomaly.id}",
            anomalyId=anomaly.id,
            profile="standard",
            rationale=(
                f"Deliberately invalid demo plan for anomaly {anomaly.id}: flies into NFZ "
                f'"{zone.name}" at {alt} m AGL, which is both inside the no-fly ceiling '
                f"({zone.ceilingM} m) and outside the site alt band "
                f"[{site.altBandMin}, {site.altBandMax}] m. The MissionVerifier must flag it."
            ),
            tools=[
                {"tool": "goto_gps", "lat": target[0], "lon": target[1],
                 "alt": alt, "profile": "standard"},
                {"tool": "rtl"},
            ],
        )


__all__ = [
    "FAILING_ALT_OVERSHOOT_M", "FAILING_ALT_UNDERSHOOT_M", "ORBIT_RADIUS_M",
    "SCRIPTED_PROFILE_MAX_ALT_M", "ScriptedPlanner", "orbit_approach_point",
]
